content/Phoenix for Rails developers - Bruno Vezoli.md

```diff
@@ -0,0 +1,347 @@
+        render(conn, "new.html", changeset: changeset)
+    end
+  end
+end
+```
+
+The main difference here is just the programming paradigm. Elixir is functional so the request (or connection) and the params are received as parameters of the action, while Ruby, being object oriented, provides the same data through local state and accessor functions. Still, the syntax is quite similar and I think it's easy to get a grasp of what's going on in the Phoenix snippet. 
+
+Another small difference is that the generator used on Phoenix will scaffold a controller that just responds to HTML, while Rails' will render views for both JSON and HTML. There is of course another generator to make a JSON resource but it'll create a new controller which will go on a different pipeline (if you go back to the routes section you'll see there's an `:api` pipeline).
+
+### Contexts
+
+Yes! My favorite part, the business logic. Remember how I said schemas on Phoenix do not call the database but are just a bag of data? Well you must have wondered where are the database calls, right here on the contexts:
+
+```elixir
+defmodule Blog.Articles do
+  @moduledoc """
+  The Articles context.
+  """
+
+  import Ecto.Query, warn: false
+  alias Blog.Repo
+
+  alias Blog.Articles.Article
+
+  def list_articles do
+    Repo.all(Article)
+  end
+
+  def get_article!(id), do: Repo.get!(Article, id)
+
+  def create_article(attrs \\ %{}) do
+    %Article{}
+    |> Article.changeset(attrs)
+    |> Repo.insert()
+  end
+
+  def update_article(%Article{} = article, attrs) do
+    article
+    |> Article.changeset(attrs)
+    |> Repo.update()
+  end
+
+  def delete_article(%Article{} = article) do
+    Repo.delete(article)
+  end
+
+  def change_article(%Article{} = article) do
+    Article.changeset(article, %{})
+  end
+end
+```
+
+Again, this follows the functional paradigm so all a context does is provide methods to interact with the data and it does so by using the repository pattern. If you are not familiar with the concept, just think of it as an abstraction of your database (or other kind of storage) where you can query, update and create records by sending the necessary data. This pattern allows to separate the data from the way it's saved, after understanding that contexts are straightforward to comprehend.
+
+This way of managing the data layer and the business logic opens up a world of oportunities in my opinion. Assume for example that we have a process for publishing in our blo, so all articles must be reviewed after they are changed and before they are published. We would probably do some validations in our model to only insert it if it's approved, same would go in our context here. But we also like to allow admins to bypass these validations, a common case, and publish whenever they want. So in Rails we would probably change our code to reflect this new case while in Phoenix we might wish to create a new context all toghether.
```

Comment:
`opportunities`
`blog`
In this case  I think it would be `altogether`, instead of `all together`.